### PR TITLE
Inverse state mapper

### PR DESCRIPTION
**Pull request by kwkbtr**

### CODE REVIEW

File: packages/chem/quri_parts/chem/transforms/__init__.py [hunk 1]
```diff
@@ -85,6 +91,28 @@ def get_state_mapper(
...
+
+        Args:
+            n_spin_orbitals:
+                The number of spin orbitals.
+            n_fermions:
+                The number of fermions considered when the qubit state is mapped.
+                Some mappings require this argument (e.g. symmetry-conserving
+                Bravyi-Kitaev transformation) while the others ignore it.
+            n_up_spins:
+                The number of spin orbitals.
```
Comment (kwkbtr) [inline packages/chem/quri_parts/chem/transforms/__init__.py]: This docstring needs to be updated.

**Review by kwkbtr — COMMENTED** [review]
Moving post-selection filter functions from `quri_parts.chem` to `quri_parts.openfermion` is a breaking change, so let's note it explicitly in pull request description.

**Comment by dchung0741** [comment]
I haven't read the post selection functions yet, but I have one question about the updated interface for incorporating the inverse mappers.
It looks like the inverse mapper could take in the `n_up_spins` argument.
For the SCBK Hamiltonian mapping, it also needs to take in `n_up_spins` in order to compute the correct parity factor for the Hamiltonians.
So my question is: Should we add the `n_up_spins` argument to the `get_state_mapper`, `get_of_operator_mapper` methods as well?

**Comment by kwkbtr** [comment]
> Should we add the `n_up_spins` argument to the `get_state_mapper`, `get_of_operator_mapper` methods as well?

I guess

- For `get_of_operator_mapper`: Yes
- For `get_state_mapper`: I think the number of up spins can be determined from the input (occupation numbers). But it is also the case for `n_fermions`, so I think it's good to add it for consistency.

File: packages/openfermion/quri_parts/openfermion/transforms/__init__.py [hunk 2]
```diff
@@ -293,5 +340,28 @@ def get_state_mapper(
...
+
+    def augment_dropped_bits(
+        self,
+        bit_array: list[int],
+        n_spin_orbitals: int,
+        n_fermions: Optional[int] = None,
+    ) -> list[int]:
+        # Add two qubits dropped by the fermion-to-qubit mapping.
+        return bit_array + [0, 0]
+
```
Comment (dchung0741) [inline packages/openfermion/quri_parts/openfermion/transforms/__init__.py]: I'm wondering if this method should be public?
My concern is that for SCBK inverse mapping, we do not rearrange the dropped qubit back to the M/2-th and M-th qubit with this method (M is the number of spin orbitals.) Would this cause user's confusions if they decide to use it as a public method?

**Review by dchung0741 — COMMENTED** [review]
The code looks good to me overall. Just one slight concern.
Also, about the `n_up_spins` should I add it in my PR later?

File: packages/openfermion/quri_parts/openfermion/transforms/__init__.py (hunk 2 above)
Comment (kwkbtr) [reply packages/openfermion/quri_parts/openfermion/transforms/__init__.py]: Yeah, maybe we can make it private (or you can call it "protected", a term used in other programming languages like C++ or Java).

Comment (toru4838) [reply packages/openfermion/quri_parts/openfermion/transforms/__init__.py]: I will make it protected in this PR

**Comment by toru4838** [comment]
Thank you for your feedback!

> Also, about the n_up_spins should I add it in my PR later?

This PR only considers restricted (OF's) transformations, so we don't need `n_up_spins` for `get_operator_mapper()` and `get_state_mapper()` here. Let's add it in your PR later.

**Review by toru4838 — APPROVED** [review]

**Pull request merged**